core/modules/link/link.module

```diff
@@ -813,30 +777,88 @@ function link_theme() {
+ *       'title' => 'A link to Content', // Prepopulates the "title" attribute.
+ *       'class' => 'additional-css-class',
+ *       'target' => '_blank',
+ *     ),
+ *   ),
+ * @endcode
  */
 function link_element_info() {
   $elements = array();
   // We use "link_field" here to avoid conflicts with theme_link().
+  // Many of these property are related to the Field API Link field settings.
   $elements['link_field'] = array(
     '#input' => TRUE,
     '#process' => array('link_field_process'),
     '#theme' => 'link_field',
     '#theme_wrappers' => array('form_element'),
+    '#element_validate' => array('link_field_element_validate'),
+
+    // Custom options for link_field. Defaults are provided within
+    // link_field_process().
+    '#link_field_options' => array(),
+
+    // #delta is typically passed by the widget form, but we provide a default
+    // value to help validations when link_field is used as a FAPI element.
+    '#delta' => 0,
+
+    // Attach the css by default.
+    '#attached' => array(
+      'css' => array(
+        backdrop_get_path('module', 'link') . '/css/link.css',
+      ),
+    ),
   );
   return $elements;
 }
 
 /**
- * Returns the default attributes and their values.
+ * Provides default values for #link_field_options properties.
  */
-function _link_default_attributes() {
+function _link_field_options_defaults() {
   return array(
+    // Possible values: 'optional' or 'required'. If 'optional', the element
+    // can be submitted without filling in the url textfield.
+    'url' => 'optional',
+
+    // Options: LINK_BOTH, LINK_INTERNAL, LINK_EXTERNAL.
+    'link_type' => LINK_BOTH,
+    'validate_url' => TRUE,
+
+    // Options: LINK_TARGET_DEFAULT, LINK_TARGET_TOP, LINK_TARGET_NEW_WINDOW
+    // and LINK_TARGET_USER. LINK_TARGET_USER will show a checkbox to indicate
+    // if the link should use '_blank' as target.
     'target' => LINK_TARGET_DEFAULT,
-    'class' => '',
-    'configurable_class' => 0,
-    'rel' => '',
-    'title' => '',
-    'configurable_title' => 0,
+    
+    // Options are: 'optional', 'required', 'value' and 'none'. If 'optional'
+    // or 'required', a textfield will be shown to enter a title for the link.
+    'title_mode' => 'optional',
+    'title_maxlength' => 128,
+
+    // If TRUE, the element's #title property will be use as the the label for
+    // the link's 'title' textfield. Otherwise, the string 'Title' will be
+    // used.
+    'title_label_use_field_label' => FALSE,
+
+    // If true, the element will show a textfield to enter the value of the
+    // "title" link attribute.
+    'configurable_title' => FALSE,
+
+    // If true, the element will show a textfield to enter a custom link
+    // class.
+    'configurable_class' => FALSE,
+
+    // When this element is used by the Field API Link field, we receive the
+    // field instance array to help us deal with special validations.
+    'instance' => array(),
```

Comment:
After all the effort to separate Field API from Form API, this seems like an undesirable thing to include. Future code could simply use the `instance` property instead of adding a new form API property. It looks like this is only used in a single place in the rest of this PR, can we eliminate it entirely?